File: Submissions/216_Combination_Sum_III.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int> > result;
        vector<int> combination;
        dfs(k, n, 1, combination, result);
        return result;
    }
    
    void dfs(int k, int n, int start, vector<int>& combination, vector<vector<int>>& result){
        if(k == 0 && n == 0) {
            result.push_back(combination);
        }else if(k <=0 || n <=0) {
            return;
        }
        for(int i=start; i<10; ++i){
            combination.push_back(i);
            dfs(k-1, n-i, i+1, combination, result);
            combination.pop_back();
        }
    }
};

File: Kotlin/216_Combination_Sum_III.kt
class Solution {
    fun combinationSum3(k: Int, n: Int): List<List<Int>> {
        val result = ArrayList<List<Int>>()
        recursive(k, n, 1, ArrayList<Int>(), result)
        return result
    }
    
    fun recursive(k: Int, n: Int, start: Int, combination: ArrayList<Int>, result: ArrayList<List<Int>>){
        if(k == 0 && n == 0){
            result.add(combination.toList())
        } else if(k <= 0 || n <= 0){
            return
        }
        for (i in start until 10){
            combination.add(i)
            recursive(k-1, n-i, i+1, combination, result)
            combination.removeAt(combination.size - 1)
        }
    }
}